src/streaming/file.sink.cpp

```diff
@@ -3,31 +3,226 @@
 
 #include <filesystem>
 
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <cstring>
+#include <fcntl.h>
+#include <sys/uio.h>
+#include <unistd.h>
+#endif
+
 namespace fs = std::filesystem;
 
-zarr::FileSink::FileSink(std::string_view filename, bool truncate)
-  : file_(filename.data(),
-          truncate ? (std::ios::binary | std::ios::trunc) : std::ios::binary)
+namespace {
+#ifdef _WIN32
+std::string
+get_last_error_as_string()
+{
+    DWORD errorMessageID = ::GetLastError();
+    if (errorMessageID == 0) {
+        return std::string(); // No error message has been recorded
+    }
+
+    LPSTR messageBuffer = nullptr;
+
+    size_t size = FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER |
+                                   FORMAT_MESSAGE_FROM_SYSTEM |
+                                   FORMAT_MESSAGE_IGNORE_INSERTS,
+                                 NULL,
+                                 errorMessageID,
+                                 MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                                 (LPSTR)&messageBuffer,
+                                 0,
+                                 NULL);
+
+    std::string message(messageBuffer, size);
+
+    LocalFree(messageBuffer);
+
+    return message;
+}
+
+void
+init_handle(void** handle, const std::string& filename)
+{
+    EXPECT(handle, "Expected nonnull pointer to file handle.");
+    HANDLE* fd = new HANDLE;
+
+    *fd = CreateFileA(filename.c_str(),
+                      GENERIC_WRITE,
+                      0, // No sharing
+                      nullptr,
+                      OPEN_ALWAYS,
+                      FILE_FLAG_OVERLAPPED,
+                      nullptr);
+
+    if (*fd == INVALID_HANDLE_VALUE) {
+        const auto err = get_last_error_as_string();
+        delete handle;
+        throw std::runtime_error("Failed to open file '" + filename +
```

Comment:
I think you can just use `std::format` now:
```cpp
throw std::runtime_error(std::format("Failed to open file {}: {}", filename, err);
```
This applies to other string formatting around errors in this PR.